### DISCUSSION TITLE
Is there any way to call Eval.PerformEval?

### DISCUSSION THREAD

**Discussion by lekrus** [sebastienros/jint#1181] [Q&A]
Hello,

Prior to beta-2036 it was possible to evaluate the script with the following code:
```c#
var engine = new Engine();
engine.Realm.Intrinsics.Eval.PerformEval("var x = 2;", engine.Realm, false, true);
```

Since introducing of ExecutionContext in 2036 it fails with `NullReferenceException` due to `_activeEvaluationContext` is null

Is there any way to achieve the same behaviour in 2036+?
We used that code to dynamically load scripts during main code execution.

Thanks!

**Comment by lahma** [comment]
So why are you not calling just `engine.Evaluate`? All direct intrinsics usage might not be ideal solution.

**Reply by lekrus** [reply]
That was the only way to set `direct = true` argument and evaluate the script in the current context and not in Global one, so when a script with `fn()` is loaded within some function, it's referenced as `this.fn()`

**Reply by lahma** [reply]
Seems that you have a very specific use case. Do you want to isolate the loaded scripts to sandbox or is `this` (what is it actually?) just logical place to attach functions?

**Reply by lekrus** [reply]
Yes, that is "legacy" inherited from early attempts to use dynamic modules before standard import/export is available.
After further investigation, I found that the `ExecutionContext` is not available only when we try to read a getter of an `ObjectInstance`
```c#
var objInstance = new ObjectInstance(engine);
CreateGetter(objInstance, "propA", "function() { dynamicLoader.Load(); return 10 + this.testFn();}");

var testObj = engine.GetValue("testObj").AsObject();
var propValue = testObj.Get("propA"); // this will throw NullRef
```
Luckily I was able to use engine.Invoke to achieve the same result:
```c#
var propGetter = testObj.GetProperty("propA").Get;
var propValue = engine.Invoke(r, testObj, Array.Empty<object>()).ToObject();
```

**Reply by lahma** [reply]
So is this legacy part something that you can rework to use newer and shinier things? I'm just trying to minimize my potential hacking work...  Building those dynamic objects outside of Jint will cause trouble down the road if they have anything special in them, like invoke needs.

**Reply by lekrus** [reply]
There is still a need for us to allow dynamic extension of the object with lazy functionality from a getter function, like in the sample above.
But if you are not going to remove `engine.Invoke` method in future Jint versions, we'll use it to achieve the goal. Thank you!
